File: HolaMundoJava/app/src/main/java/com/example/holamundojava/backend/Grafico.java
package com.example.holamundojava.backend;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.Log;
import android.view.View;

import androidx.annotation.NonNull;

public class Grafico extends View {
    private static final String TAG = "GRAFICO";

    String tipoGrafico;
    public Grafico(Context context, String tipoGrafico) {
        super(context);
        this.tipoGrafico = tipoGrafico;
    }

    @Override
    protected void onDraw(@NonNull Canvas canvas) {
        super.onDraw(canvas);
        canvas.drawColor(Color.argb(255, 125, 100, 180));
        Paint paint = new Paint();

        if (tipoGrafico.equals("linea")) {
            drawLineas(canvas, paint);
        } else if (tipoGrafico.equals("circulo")) {
            drawCirculos(canvas, paint);
        } else if (tipoGrafico.equals("ramdonCirculo")) {
            ramdonCircles(canvas, paint);
        } else if (tipoGrafico.equals("arbol")) {
            drawArbol(canvas, paint);
        } else if (tipoGrafico.equals("2D")) {
            planoCartesiano2D(canvas, paint);
        } else if (tipoGrafico.equals("3D")) {
            planoCartesiano3D(canvas, paint);
        }
    }
    private void planoCartesiano3D(Canvas canvas, Paint paint) {
        canvas.drawColor(Color.BLACK);
        paint.setColor(Color.WHITE);
        paint.setStrokeWidth(5);
        int ancho = getWidth();
        int alto = getHeight();
        // dibjuar los 3 ejes
        canvas.drawLine(0, alto/2, ancho, alto/2, paint);
        canvas.drawLine(ancho/2, 0, ancho/2, alto, paint);
        canvas.drawLine(ancho/2-500, alto/2+500, ancho/2+500, alto/2-500, paint);
        paint.setColor(Color.YELLOW);

        for(float j=0; j<=10; j+=0.03){
            float limitInfX = -20+j;
            float limitSupX = 20+j;
            float LimitInfY = -20+j;
            float LimitSupY = 20+j;
            for (float x = limitInfX; x <= limitSupX; x += 0.1) {
                double Y = fx(x);
                double xt = ((x - limitInfX) / (limitSupX - limitInfX)) * ancho;
                double yt = alto - ((Y - LimitInfY) / (LimitSupY - LimitInfY)) * alto;
                //ouble zt = ((x - limitInfZ) / (limitSupZ - limitInfZ)) * ancho;
                paint.setColor(Color.YELLOW);
                canvas.drawCircle((float) xt, (float) yt, 3, paint);
            }
        }
    }
    private void planoCartesiano2D(Canvas canvas, Paint paint) {
        canvas.drawColor(Color.BLACK);
        paint.setColor(Color.WHITE);
        paint.setStrokeWidth(5);
        int ancho = getWidth();
        int alto = getHeight();
        Log.d(TAG, "ancho: " + ancho + " alto: " + alto);
        canvas.drawLine(0, alto/2, ancho, alto/2, paint);
        canvas.drawLine(ancho/2, 0, ancho/2, alto, paint);
        float limInfX = -20;
        float limSupX = 20;
        float limInfY = -20;
        float limSupY = 20;
        paint.setColor(Color.YELLOW);
        // f de (x) = x^2
        for (float x = limInfX; x <= limSupX; x+=0.01) {
            float y = fx(x);
            //Log.d(TAG, "x: " + x + " y: " + y);
            float xt = (x-limInfX)/(limSupX-limInfX)*ancho;
            float yt = alto - (((y-limInfY)/(limSupY-limInfY))*alto);
            canvas.drawCircle(xt, yt, 3, paint);
            Log.d(TAG, "x"+x+"xt: " + xt + " y"+y+"yt: " + yt);
        }
    }
    private float fx(float x) {
        return x*x;
    }
    private void ramdonCircles(Canvas canvas, Paint paint) {
        for (int i = 0; i < 200; i++) {
            int x = (int) (Math.random() * 1000);
            int y = (int) (Math.random() * 2000);
            int r = (int) (Math.random() * 100);
            paint.setColor(Color.argb(100, 10*i+5, 5*i+10, 3*i+20));
            canvas.drawCircle(x, y, r, paint);
        }
    }
    private void drawCirculos(Canvas canvas, Paint paint) {
        for (int i = 0; i < 200; i++) {
            paint.setColor(Color.argb(100, 10*i+5, 5*i+10, 3*i+20));
            canvas.drawCircle(200+i*3, 1000+i*2, 150, paint);
        }
    }
    private void drawLineas(Canvas canvas, Paint paint) {
        for (int i = 0; i < 200; i++) {
            paint.setColor(Color.argb(100, 10*i+5, 5*i+10, 3*i+20));
            canvas.drawLine(10, 10, 600-i*3, 3*i, paint);
        }
    }
    private void drawArbol(Canvas canvas, Paint paint) {
        canvas.drawColor(Color.WHITE);
        paint.setColor(Color.argb(255, 0, 255, 0));
        // hojas del arbol
        for (int i = 0; i < 320; i++) {
            canvas.drawLine(500, 300, 1000-i*3, 1300, paint);
        }
        // tronco del arbol
        for (int i = 0; i < 170; i++) {
            paint.setColor(Color.argb(200, 100, 50, 0));
            canvas.drawLine(450+i, 1300, 450+i, 1700, paint);
        }
        // focos de luz del arbol
        ramdonCirculosArbol(canvas, paint);

    }
    private void ramdonCirculosArbol(Canvas canvas, Paint paint) {
        // Puntos del triángulo
        int x1 = 500, y1 = 300;
        int x2 = 50, y2 = 1300;
        int x3 = 1000, y3 = 1300;

        // Generar círculos aleatorios dentro del triángulo
        for (int i = 0; i < 150; i++) {
            paint.setColor(Color.argb(100, 10*i+5, 5*i+10, 3*i+20));
            // Generar coordenadas aleatorias dentro del triángulo
            double r1 = Math.random();
            double r2 = Math.random();

            // Asegurarse de que los puntos estén dentro del triángulo
            if ((r1 + r2) <= 1) {
                int x = (int) (x1 + r1 * (x2 - x1) + r2 * (x3 - x1));
                int y = (int) (y1 + r1 * (y2 - y1) + r2 * (y3 - y1));
                int radius = (int) (Math.random() * 70);
                canvas.drawCircle(x, y, radius, paint);
            }
        }
    }
}

File: HolaMundoKotlin/app/src/main/java/com/example/holamundokotlin/backend/Grafico.kt
package com.example.holamundokotlin.backend

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.LinearGradient
import android.graphics.Paint
import android.graphics.Shader
import android.view.View


class Grafico(context: Context?, var tipoGrafico: String) : View(context) {
    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        canvas.drawColor(Color.argb(255, 125, 100, 180))
        val paint = Paint()
        if (tipoGrafico == "cuadrado") {
            drawCuadrado(canvas, paint, 500.0f)
        } else if(tipoGrafico == "cubo"){
            drawCubo(canvas, paint, 500.0f)
        } else if (tipoGrafico == "2D") {
            planoCartesiano2D(canvas, paint)
        } else if (tipoGrafico == "3D") {
            planoCartesiano3D(canvas, paint)
        } else if (tipoGrafico == "caratula"){
            drawCaratula(canvas, paint);
        }
    }

    private fun drawCaratula(canvas: Canvas, paint: Paint) {
        // Fondo gradiente
        val gradient = LinearGradient(
            0f, 0f, canvas.width.toFloat(), canvas.height.toFloat(),
            Color.rgb(0, 128, 128), Color.rgb(0, 0, 255),
            Shader.TileMode.CLAMP
        )
        paint.shader = gradient
        canvas.drawRect(0f, 0f, canvas.width.toFloat(), canvas.height.toFloat(), paint)

        // Dibujar patrón en la parte superior
        paint.shader = null
        paint.color = Color.YELLOW
        for (i in 0 until canvas.width step 40) {
            canvas.drawLine(i.toFloat(), 0f, (i + 20).toFloat(), 100f, paint)
            canvas.drawLine((i + 20).toFloat(), 100f, (i + 40).toFloat(), 0f, paint)
        }

        // Dibujar patrón en la parte inferior
        paint.color = Color.MAGENTA
        for (i in 0 until canvas.width step 40) {
            canvas.drawLine(i.toFloat(), canvas.height.toFloat(), (i + 20).toFloat(), canvas.height.toFloat() - 100f, paint)
            canvas.drawLine((i + 20).toFloat(), canvas.height.toFloat() - 100f, (i + 40).toFloat(), canvas.height.toFloat(), paint)
        }

        paint.setColor(Color.YELLOW)
        paint.strokeWidth = 3f
        for (i in 0..17) {
            canvas.drawLine(200f, 550f, 900f-20f*i, (550 +20*i).toFloat(), paint)
            canvas.drawLine(900f, 550f, 200f+20f*i, (550 +20*i).toFloat(), paint)
        }
        val centerX = canvas.width / 2f + 50f
        val centerY = 1500f
        var lado = 400f
        val halfLado = lado / 2f
        paint.setColor(Color.MAGENTA)
        // Establecer el grosor de la línea y el color
        paint.strokeWidth = 5f  // Grosor de la línea

        // Coordenadas de los vértices del cubo
        val vertices = arrayOf(
            floatArrayOf(centerX - halfLado, centerY - halfLado),  // Vértice 0 (frontal superior izquierdo)
            floatArrayOf(centerX + halfLado, centerY - halfLado),  // Vértice 1 (frontal superior derecho)
            floatArrayOf(centerX + halfLado, centerY + halfLado),  // Vértice 2 (frontal inferior derecho)
            floatArrayOf(centerX - halfLado, centerY + halfLado),  // Vértice 3 (frontal inferior izquierdo)
            floatArrayOf(centerX - halfLado - lado / 4, centerY - halfLado - lado / 4),  // Vértice 4 (trasero superior izquierdo)
            floatArrayOf(centerX + halfLado - lado / 4, centerY - halfLado - lado / 4),  // Vértice 5 (trasero superior derecho)
            floatArrayOf(centerX + halfLado - lado / 4, centerY + halfLado - lado / 4),  // Vértice 6 (trasero inferior derecho)
            floatArrayOf(centerX - halfLado - lado / 4, centerY + halfLado - lado / 4)   // Vértice 7 (trasero inferior izquierdo)
        )

        // Dibujar líneas para formar las aristas del cubo
        canvas.drawLine(vertices[0][0], vertices[0][1], vertices[1][0], vertices[1][1], paint) // Arista frontal superior
        canvas.drawLine(vertices[1][0], vertices[1][1], vertices[2][0], vertices[2][1], paint) // Arista frontal derecha
        canvas.drawLine(vertices[2][0], vertices[2][1], vertices[3][0], vertices[3][1], paint) // Arista frontal inferior
        canvas.drawLine(vertices[3][0], vertices[3][1], vertices[0][0], vertices[0][1], paint) // Arista frontal izquierda

        canvas.drawLine(vertices[4][0], vertices[4][1], vertices[5][0], vertices[5][1], paint) // Arista trasera superior
        canvas.drawLine(vertices[5][0], vertices[5][1], vertices[6][0], vertices[6][1], paint) // Arista trasera derecha
        canvas.drawLine(vertices[6][0], vertices[6][1], vertices[7][0], vertices[7][1], paint) // Arista trasera inferior
        canvas.drawLine(vertices[7][0], vertices[7][1], vertices[4][0], vertices[4][1], paint) // Arista trasera izquierda

        canvas.drawLine(vertices[0][0], vertices[0][1], vertices[4][0], vertices[4][1], paint) // Arista superior izquierda
        canvas.drawLine(vertices[1][0], vertices[1][1], vertices[5][0], vertices[5][1], paint) // Arista superior derecha
        canvas.drawLine(vertices[2][0], vertices[2][1], vertices[6][0], vertices[6][1], paint) // Arista inferior derecha
        canvas.drawLine(vertices[3][0], vertices[3][1], vertices[7][0], vertices[7][1], paint)

        // Configurar el color y el tamaño del texto
        paint.color = Color.MAGENTA
        paint.textSize = 120f

        // Dibujar el texto "SIS104"
        val textSIS104 = "SIS104"
        val xSIS104 = canvas.width / 2f - 140f
        val ySIS104 = 500f
        canvas.drawText(textSIS104, xSIS104, ySIS104, paint)

        // Dibujar el texto "MAAI"
        paint.color = Color.YELLOW
        val textMAAI = "MAAI"
        val xMAAI = canvas.width / 2f - 90f
        val yMAAI = 1500f
        canvas.drawText(textMAAI, xMAAI, yMAAI, paint)
    }

    private fun planoCartesiano3D(canvas: Canvas, paint: Paint) {
        canvas.drawColor(Color.BLACK)
        paint.setColor(Color.WHITE)
        paint.strokeWidth = 5f
        val ancho = width
        val alto = height
        // dibjuar los 3 ejes
        canvas.drawLine(0f, (alto / 2).toFloat(), ancho.toFloat(), (alto / 2).toFloat(), paint)
        canvas.drawLine((ancho / 2).toFloat(), 0f, (ancho / 2).toFloat(), alto.toFloat(), paint)
        canvas.drawLine(
            (ancho / 2 - 500).toFloat(),
            (alto / 2 + 500).toFloat(),
            (ancho / 2 + 500).toFloat(),
            (alto / 2 - 500).toFloat(),
            paint
        )
        paint.setColor(Color.YELLOW)
        var j = 0f
        while (j <= 10) {
            val limitInfX = -20 + j
            val limitSupX = 20 + j
            val LimitInfY = -20 + j
            val LimitSupY = 20 + j
            var x = limitInfX
            while (x <= limitSupX) {
                val Y = fx(x).toDouble()
                val xt = ((x - limitInfX) / (limitSupX - limitInfX) * ancho).toDouble()
                val yt = alto - (Y - LimitInfY) / (LimitSupY - LimitInfY) * alto
                //ouble zt = ((x - limitInfZ) / (limitSupZ - limitInfZ)) * ancho;
                paint.setColor(Color.YELLOW)
                canvas.drawCircle(xt.toFloat(), yt.toFloat(), 3f, paint)
                x += 0.1.toFloat()
            }
            j += 0.03.toFloat()
        }
    }

    private fun planoCartesiano2D(canvas: Canvas, paint: Paint) {
        canvas.drawColor(Color.BLACK)
        paint.setColor(Color.WHITE)
        paint.strokeWidth = 5f
        val ancho = width
        val alto = height
        canvas.drawLine(0f, (alto / 2).toFloat(), ancho.toFloat(), (alto / 2).toFloat(), paint)
        canvas.drawLine((ancho / 2).toFloat(), 0f, (ancho / 2).toFloat(), alto.toFloat(), paint)
        val limInfX = -20f
        val limSupX = 20f
        val limInfY = -20f
        val limSupY = 20f
        paint.setColor(Color.YELLOW)
        // f de (x) = x^2
        var x = limInfX
        while (x <= limSupX) {
            val y = fx(x)
            //Log.d(TAG, "x: " + x + " y: " + y);
            val xt = (x - limInfX) / (limSupX - limInfX) * ancho
            val yt = alto - (y - limInfY) / (limSupY - limInfY) * alto
            canvas.drawCircle(xt, yt, 3f, paint)
            x += 0.01.toFloat()
        }
    }

    private fun fx(x: Float): Float {
        return x * x
    }
    private fun drawCubo(canvas: Canvas, paint: Paint, ladoA: Float) {
        val centerX = canvas.width / 2f
        val centerY = canvas.height / 2f
        val screenWidth = canvas.width.toFloat()
        var lado = 0.0f
        if (ladoA > screenWidth){
            lado = calcularLadoEscalado(screenWidth, ladoA)
        } else {
            lado = ladoA
        }
        val halfLado = lado / 2f

        // Establecer el grosor de la línea y el color
        paint.strokeWidth = 5f  // Grosor de la línea
        paint.color = Color.CYAN  // Color de la línea

        // Coordenadas de los vértices del cubo
        val vertices = arrayOf(
            floatArrayOf(centerX - halfLado, centerY - halfLado),  // Vértice 0 (frontal superior izquierdo)
            floatArrayOf(centerX + halfLado, centerY - halfLado),  // Vértice 1 (frontal superior derecho)
            floatArrayOf(centerX + halfLado, centerY + halfLado),  // Vértice 2 (frontal inferior derecho)
            floatArrayOf(centerX - halfLado, centerY + halfLado),  // Vértice 3 (frontal inferior izquierdo)
            floatArrayOf(centerX - halfLado - lado / 4, centerY - halfLado - lado / 4),  // Vértice 4 (trasero superior izquierdo)
            floatArrayOf(centerX + halfLado - lado / 4, centerY - halfLado - lado / 4),  // Vértice 5 (trasero superior derecho)
            floatArrayOf(centerX + halfLado - lado / 4, centerY + halfLado - lado / 4),  // Vértice 6 (trasero inferior derecho)
            floatArrayOf(centerX - halfLado - lado / 4, centerY + halfLado - lado / 4)   // Vértice 7 (trasero inferior izquierdo)
        )

        // Dibujar líneas para formar las aristas del cubo
        canvas.drawLine(vertices[0][0], vertices[0][1], vertices[1][0], vertices[1][1], paint) // Arista frontal superior
        canvas.drawLine(vertices[1][0], vertices[1][1], vertices[2][0], vertices[2][1], paint) // Arista frontal derecha
        canvas.drawLine(vertices[2][0], vertices[2][1], vertices[3][0], vertices[3][1], paint) // Arista frontal inferior
        canvas.drawLine(vertices[3][0], vertices[3][1], vertices[0][0], vertices[0][1], paint) // Arista frontal izquierda

        canvas.drawLine(vertices[4][0], vertices[4][1], vertices[5][0], vertices[5][1], paint) // Arista trasera superior
        canvas.drawLine(vertices[5][0], vertices[5][1], vertices[6][0], vertices[6][1], paint) // Arista trasera derecha
        canvas.drawLine(vertices[6][0], vertices[6][1], vertices[7][0], vertices[7][1], paint) // Arista trasera inferior
        canvas.drawLine(vertices[7][0], vertices[7][1], vertices[4][0], vertices[4][1], paint) // Arista trasera izquierda

        canvas.drawLine(vertices[0][0], vertices[0][1], vertices[4][0], vertices[4][1], paint) // Arista superior izquierda
        canvas.drawLine(vertices[1][0], vertices[1][1], vertices[5][0], vertices[5][1], paint) // Arista superior derecha
        canvas.drawLine(vertices[2][0], vertices[2][1], vertices[6][0], vertices[6][1], paint) // Arista inferior derecha
        canvas.drawLine(vertices[3][0], vertices[3][1], vertices[7][0], vertices[7][1], paint) // Arista inferior izquierda
    }


    private fun drawCuadrado(canvas: Canvas, paint: Paint, lado: Float) {
        paint.setColor(Color.CYAN)
        var ladoEscalado = 1.0f
        val screenWidth = canvas.width.toFloat()
        val screenHeight = canvas.height.toFloat()

        if (lado > screenWidth){
            ladoEscalado = calcularLadoEscalado(screenWidth, lado)
        } else {
            ladoEscalado = lado
        }

        val centerX = screenWidth / 2
        val centerY = screenHeight / 2

        val left = centerX - ladoEscalado / 2
        val top = centerY - ladoEscalado / 2
        val right = centerX + ladoEscalado / 2
        val bottom = centerY + ladoEscalado / 2

        canvas.drawRect(left, top, right, bottom, paint)
    }
    private fun calcularLadoEscalado(anchoPantalla: Float, lado: Float): Float {
        return if (lado >= 100000.0f) {
            anchoPantalla
        } else {
            lado * (anchoPantalla / 100000f)
        }
    }
}